In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression

In [2]:
clf = LogisticRegression()

In [3]:
df = pd.read_csv("0000000000002429_training_titanic_x_y_train.csv")

In [4]:
df.head()

Pclass                                            Name     Sex   Age  \
0       2   Weisz, Mrs. Leopold (Mathilde Francoise Pede)  female  29.0   
1       3               Williams, Mr. Howard Hugh "Harry"    male   NaN   
2       2  Morley, Mr. Henry Samuel ("Mr Henry Marshall")    male  39.0   
3       3     Palsson, Mrs. Nils (Alma Cornelia Berglund)  female  29.0   
4       3                          Sutehall, Mr. Henry Jr    male  25.0   

   SibSp  Parch           Ticket    Fare Cabin Embarked  Survived  
0      1      0           228414  26.000   NaN        S         1  
1      0      0         A/5 2466   8.050   NaN        S         0  
2      0      0           250655  26.000   NaN        S         0  
3      0      4           349909  21.075   NaN        S         0  
4      0      0  SOTON/OQ 392076   7.050   NaN        S         0

Name and Ticket Column don't seem useful,as passenger's survival doesn't depend on name, so it's better to drop them

In [5]:
del df["Name"]
del df["Ticket"]

In [6]:
df.head()

Pclass     Sex   Age  SibSp  Parch    Fare Cabin Embarked  Survived
0       2  female  29.0      1      0  26.000   NaN        S         1
1       3    male   NaN      0      0   8.050   NaN        S         0
2       2    male  39.0      0      0  26.000   NaN        S         0
3       3  female  29.0      0      4  21.075   NaN        S         0
4       3    male  25.0      0      0   7.050   NaN        S         0

In [7]:
df.isnull().sum()

Pclass        0
Sex           0
Age         132
SibSp         0
Parch         0
Fare          0
Cabin       514
Embarked      1
Survived      0
dtype: int64

As there are 514 null values in Cabin column, it's better to drop it, Also dropping fare column

In [8]:
del df["Cabin"]
del df["Fare"]

In [9]:
df.head(20)

Pclass     Sex   Age  SibSp  Parch Embarked  Survived
0        2  female  29.0      1      0        S         1
1        3    male   NaN      0      0        S         0
2        2    male  39.0      0      0        S         0
3        3  female  29.0      0      4        S         0
4        3    male  25.0      0      0        S         0
5        3    male  34.5      0      0        C         0
6        1  female  35.0      1      0        S         1
7        3    male  23.0      0      0        S         0
8        2    male   NaN      0      0        S         1
9        3    male  50.0      0      0        S         0
10       3  female  23.0      0      0        S         1
11       3    male   NaN      0      0        C         0
12       2    male  36.0      0      0        S         0
13       3    male   NaN      1      0        Q         0
14       3  female   NaN      0      0        Q         1
15       3    male  28.0      2      0        S         0
16       3    male  32.0      0      0        S         1
17       3    male  42.0      0      0        S         0
18       1    male  38.0      0      1        S         0
19       3  female   NaN      0      0        Q         1

In [10]:
mean_age = df["Age"].mean()
mean_age

29.700559701492537

In [11]:
df.Age.fillna(mean_age,inplace=True)
df

Pclass     Sex       Age  SibSp  Parch Embarked  Survived
0         2  female  29.00000      1      0        S         1
1         3    male  29.70056      0      0        S         0
2         2    male  39.00000      0      0        S         0
3         3  female  29.00000      0      4        S         0
4         3    male  25.00000      0      0        S         0
..      ...     ...       ...    ...    ...      ...       ...
663       2  female  17.00000      0      0        S         1
664       3    male  29.70056      0      0        Q         0
665       3    male  32.00000      0      0        S         1
666       3  female  22.00000      0      0        S         0
667       3  female  29.70056      1      0        Q         1

[668 rows x 7 columns]

In [12]:
df.isnull().sum()

Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Embarked    1
Survived    0
dtype: int64

In [13]:
df.dropna(inplace=True)
df.isnull().sum()

Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Embarked    0
Survived    0
dtype: int64

Converting Male and Female into numerical values. Male = 1, Female = 2

In [14]:
def getNumber(str):
    if str=="male":
        return 1
    else:
        return 2
df["gender"]= df["Sex"].apply(getNumber)

In [15]:
df.head(10)

Pclass     Sex       Age  SibSp  Parch Embarked  Survived  gender
0       2  female  29.00000      1      0        S         1       2
1       3    male  29.70056      0      0        S         0       1
2       2    male  39.00000      0      0        S         0       1
3       3  female  29.00000      0      4        S         0       2
4       3    male  25.00000      0      0        S         0       1
5       3    male  34.50000      0      0        C         0       1
6       1  female  35.00000      1      0        S         1       2
7       3    male  23.00000      0      0        S         0       1
8       2    male  29.70056      0      0        S         1       1
9       3    male  50.00000      0      0        S         0       1

In [16]:
del df["Sex"]
df.head(10)

Pclass       Age  SibSp  Parch Embarked  Survived  gender
0       2  29.00000      1      0        S         1       2
1       3  29.70056      0      0        S         0       1
2       2  39.00000      0      0        S         0       1
3       3  29.00000      0      4        S         0       2
4       3  25.00000      0      0        S         0       1
5       3  34.50000      0      0        C         0       1
6       1  35.00000      1      0        S         1       2
7       3  23.00000      0      0        S         0       1
8       2  29.70056      0      0        S         1       1
9       3  50.00000      0      0        S         0       1

In [17]:
def getNumber2(str):
    if str=="S":
        return 1
    elif str=="C":
        return 2
    else :
        return 3

In [18]:
df["embark"]= df["Embarked"].apply(getNumber2)
df

Pclass       Age  SibSp  Parch Embarked  Survived  gender  embark
0         2  29.00000      1      0        S         1       2       1
1         3  29.70056      0      0        S         0       1       1
2         2  39.00000      0      0        S         0       1       1
3         3  29.00000      0      4        S         0       2       1
4         3  25.00000      0      0        S         0       1       1
..      ...       ...    ...    ...      ...       ...     ...     ...
663       2  17.00000      0      0        S         1       2       1
664       3  29.70056      0      0        Q         0       1       3
665       3  32.00000      0      0        S         1       1       1
666       3  22.00000      0      0        S         0       2       1
667       3  29.70056      1      0        Q         1       2       3

[667 rows x 8 columns]

In [19]:
df.head(20)

Pclass       Age  SibSp  Parch Embarked  Survived  gender  embark
0        2  29.00000      1      0        S         1       2       1
1        3  29.70056      0      0        S         0       1       1
2        2  39.00000      0      0        S         0       1       1
3        3  29.00000      0      4        S         0       2       1
4        3  25.00000      0      0        S         0       1       1
5        3  34.50000      0      0        C         0       1       2
6        1  35.00000      1      0        S         1       2       1
7        3  23.00000      0      0        S         0       1       1
8        2  29.70056      0      0        S         1       1       1
9        3  50.00000      0      0        S         0       1       1
10       3  23.00000      0      0        S         1       2       1
11       3  29.70056      0      0        C         0       1       2
12       2  36.00000      0      0        S         0       1       1
13       3  29.70056      1      0        Q         0       1       3
14       3  29.70056      0      0        Q         1       2       3
15       3  28.00000      2      0        S         0       1       1
16       3  32.00000      0      0        S         1       1       1
17       3  42.00000      0      0        S         0       1       1
18       1  38.00000      0      1        S         0       1       1
19       3  29.70056      0      0        Q         1       2       3

In [20]:
del df["Embarked"]
df.head(10)

Pclass       Age  SibSp  Parch  Survived  gender  embark
0       2  29.00000      1      0         1       2       1
1       3  29.70056      0      0         0       1       1
2       2  39.00000      0      0         0       1       1
3       3  29.00000      0      4         0       2       1
4       3  25.00000      0      0         0       1       1
5       3  34.50000      0      0         0       1       2
6       1  35.00000      1      0         1       2       1
7       3  23.00000      0      0         0       1       1
8       2  29.70056      0      0         1       1       1
9       3  50.00000      0      0         0       1       1

In [21]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()

In [22]:
df[["Age"]] = scaler.fit_transform(df[["Age"]])

In [23]:
df

Pclass       Age  SibSp  Parch  Survived  gender  embark
0         2  0.357116      1      0         1       2       1
1         3  0.365947      0      0         0       1       1
2         2  0.483172      0      0         0       1       1
3         3  0.357116      0      4         0       2       1
4         3  0.306694      0      0         0       1       1
..      ...       ...    ...    ...       ...     ...     ...
663       2  0.205849      0      0         1       2       1
664       3  0.365947      0      0         0       1       3
665       3  0.394933      0      0         1       1       1
666       3  0.268877      0      0         0       2       1
667       3  0.365947      1      0         1       2       3

[667 rows x 7 columns]

In [24]:
df["Pclass"].unique()

array([2, 3, 1], dtype=int64)

In [25]:
df.shape

(667, 7)

In [26]:
x = df.drop(["Survived"],axis=1)
X = np.array(x)
X

array([[2.        , 0.35711585, 1.        , 0.        , 2.        ,
        1.        ],
       [3.        , 0.3659468 , 0.        , 0.        , 1.        ,
        1.        ],
       [2.        , 0.48317156, 0.        , 0.        , 1.        ,
        1.        ],
       ...,
       [3.        , 0.39493256, 0.        , 0.        , 1.        ,
        1.        ],
       [3.        , 0.26887684, 0.        , 0.        , 2.        ,
        1.        ],
       [3.        , 0.3659468 , 1.        , 0.        , 2.        ,
        3.        ]])

In [27]:
y=df["Survived"]
Y=np.array(y)
Y

array([1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0,
       1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1,
       0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0,
       0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1,
       0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0,
       1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1,
       1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0,
       1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1,
       1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1,

In [28]:
clf.fit(X, Y)

LogisticRegression()

In [29]:
df_test = pd.read_csv("0000000000002429_test_titanic_x_test.csv")
df_test.head()

Pclass                                               Name     Sex   Age  \
0       2                     Davies, Master. John Morgan Jr    male   8.0   
1       1                        Leader, Dr. Alice (Farnham)  female  49.0   
2       3                            Kilgannon, Mr. Thomas J    male   NaN   
3       2  Jacobsohn, Mrs. Sidney Samuel (Amy Frances Chr...  female  24.0   
4       1                          McGough, Mr. James Robert    male  36.0   

   SibSp  Parch      Ticket     Fare Cabin Embarked  
0      1      1  C.A. 33112  36.7500   NaN        S  
1      0      0       17465  25.9292   D17        S  
2      0      0       36865   7.7375   NaN        Q  
3      2      1      243847  27.0000   NaN        S  
4      0      0    PC 17473  26.2875   E25        S

In [30]:
df_test.isnull().sum()

Pclass        0
Name          0
Sex           0
Age          45
SibSp         0
Parch         0
Ticket        0
Fare          0
Cabin       173
Embarked      1
dtype: int64

In [31]:
del df_test["Name"]
del df_test["Ticket"]
del df_test["Fare"]
del df_test["Cabin"]
df_test

Pclass     Sex   Age  SibSp  Parch Embarked
0         2    male   8.0      1      1        S
1         1  female  49.0      0      0        S
2         3    male   NaN      0      0        Q
3         2  female  24.0      2      1        S
4         1    male  36.0      0      0        S
..      ...     ...   ...    ...    ...      ...
218       3    male  20.0      1      0        S
219       1    male  45.0      0      0        S
220       1  female  17.0      1      0        C
221       3    male  43.0      0      0        S
222       2    male  36.5      0      2        S

[223 rows x 6 columns]

In [32]:
mean_age_test = df_test["Age"].mean()
df_test.Age.fillna(mean_age_test,inplace=True)
df_test

Pclass     Sex        Age  SibSp  Parch Embarked
0         2    male   8.000000      1      1        S
1         1  female  49.000000      0      0        S
2         3    male  29.694775      0      0        Q
3         2  female  24.000000      2      1        S
4         1    male  36.000000      0      0        S
..      ...     ...        ...    ...    ...      ...
218       3    male  20.000000      1      0        S
219       1    male  45.000000      0      0        S
220       1  female  17.000000      1      0        C
221       3    male  43.000000      0      0        S
222       2    male  36.500000      0      2        S

[223 rows x 6 columns]

In [33]:
df_test.isnull().sum()

Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Embarked    1
dtype: int64

In [34]:
df.dropna(inplace=True)

In [35]:
df_test["gender"]= df_test["Sex"].apply(getNumber)
df_test["embark"]= df_test["Embarked"].apply(getNumber2)
df_test

Pclass     Sex        Age  SibSp  Parch Embarked  gender  embark
0         2    male   8.000000      1      1        S       1       1
1         1  female  49.000000      0      0        S       2       1
2         3    male  29.694775      0      0        Q       1       3
3         2  female  24.000000      2      1        S       2       1
4         1    male  36.000000      0      0        S       1       1
..      ...     ...        ...    ...    ...      ...     ...     ...
218       3    male  20.000000      1      0        S       1       1
219       1    male  45.000000      0      0        S       1       1
220       1  female  17.000000      1      0        C       2       2
221       3    male  43.000000      0      0        S       1       1
222       2    male  36.500000      0      2        S       1       1

[223 rows x 8 columns]

In [36]:
del df_test["Sex"]
del df_test["Embarked"]
df_test

Pclass        Age  SibSp  Parch  gender  embark
0         2   8.000000      1      1       1       1
1         1  49.000000      0      0       2       1
2         3  29.694775      0      0       1       3
3         2  24.000000      2      1       2       1
4         1  36.000000      0      0       1       1
..      ...        ...    ...    ...     ...     ...
218       3  20.000000      1      0       1       1
219       1  45.000000      0      0       1       1
220       1  17.000000      1      0       2       2
221       3  43.000000      0      0       1       1
222       2  36.500000      0      2       1       1

[223 rows x 6 columns]

In [37]:
df_test.isnull().sum()

Pclass    0
Age       0
SibSp     0
Parch     0
gender    0
embark    0
dtype: int64

In [38]:
scaler2 = MinMaxScaler()

In [39]:
df_test[["Age"]] = scaler2.fit_transform(df_test[["Age"]])

In [40]:
df_test

Pclass       Age  SibSp  Parch  gender  embark
0         2  0.107396      1      1       1       1
1         1  0.688297      0      0       2       1
2         3  0.414774      0      0       1       3
3         2  0.334089      2      1       2       1
4         1  0.504109      0      0       1       1
..      ...       ...    ...    ...     ...     ...
218       3  0.277416      1      0       1       1
219       1  0.631624      0      0       1       1
220       1  0.234911      1      0       2       2
221       3  0.603287      0      0       1       1
222       2  0.511193      0      2       1       1

[223 rows x 6 columns]

In [41]:
x.head(5)

Pclass       Age  SibSp  Parch  gender  embark
0       2  0.357116      1      0       2       1
1       3  0.365947      0      0       1       1
2       2  0.483172      0      0       1       1
3       3  0.357116      0      4       2       1
4       3  0.306694      0      0       1       1

In [42]:
df_test.describe()

Pclass         Age       SibSp       Parch      gender      embark
count  223.000000  223.000000  223.000000  223.000000  223.000000  223.000000
mean     2.345291    0.414774    0.506726    0.304933    1.327354    1.408072
std      0.850047    0.194802    1.169700    0.634108    0.470303    0.703484
min      1.000000    0.000000    0.000000    0.000000    1.000000    1.000000
25%      2.000000    0.305752    0.000000    0.000000    1.000000    1.000000
50%      3.000000    0.414774    0.000000    0.000000    1.000000    1.000000
75%      3.000000    0.489940    1.000000    0.000000    2.000000    2.000000
max      3.000000    1.000000    8.000000    2.000000    2.000000    3.000000

In [43]:
df_test.isnull().sum()

Pclass    0
Age       0
SibSp     0
Parch     0
gender    0
embark    0
dtype: int64

In [44]:
X_test = np.array(df_test)
X_test

array([[2.        , 0.10739586, 1.        , 1.        , 1.        ,
        1.        ],
       [1.        , 0.68829697, 0.        , 0.        , 2.        ,
        1.        ],
       [3.        , 0.41477437, 0.        , 0.        , 1.        ,
        3.        ],
       ...,
       [1.        , 0.23491074, 1.        , 0.        , 2.        ,
        2.        ],
       [3.        , 0.60328705, 0.        , 0.        , 1.        ,
        1.        ],
       [2.        , 0.51119297, 0.        , 2.        , 1.        ,
        1.        ]])

In [45]:
Y_pred = clf.predict(X_test)
Y_pred

array([0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0,
       1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0,
       1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0,
       0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0,
       0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0,
       1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0,
       0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1,
       1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0,
       1, 0, 0], dtype=int64)

In [46]:
#np.savetxt('Y_Predictions_Titanic Assignment2.csv', Y_pred, fmt = '%d')

In [47]:
#result = np.loadtxt("Y_Predictions_Titanic Assignment2.csv")
#result

In [49]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.tree import export_graphviz
import pydotplus
from sklearn.metrics import confusion_matrix

In [50]:
clf = DecisionTreeClassifier()
clf.fit(X, Y)

DecisionTreeClassifier()

In [52]:
import pydotplus
dot_data = export_graphviz(clf, out_file=None)
graph = pydotplus.graph_from_dot_data(dot_data)
graph.write_pdf("titanic.pdf")

True

In [53]:
y_train_pred = clf.predict(X)

In [54]:
confusion_matrix(Y,y_train_pred )

array([[393,   6],
       [ 33, 235]], dtype=int64)